### PR TITLE
Only apply hostedcluster updates when the resource has changed

### PR DESCRIPTION
**Pull request by ironcladlou**
Before this commit, hostedcluster controller would perform client update
calls whether or not there was any change to the resource, resulting in
excessive no-op calls to the apiserver.

This commit adds semantic equality checks around the updates so that the
API calls are only made when the resource has changes to apply.

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Comment by ironcladlou** [comment]
Instead I think I'll refactor all the direct client calls with createOrUpdateProvider

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 1]
```diff
@@ -4262,8 +4276,10 @@ func (r *HostedClusterReconciler) defaultClusterIDsIfNeeded(ctx context.Context,
 	}
 
 	if needsUpdate {
-		if err := r.Update(ctx, hcluster); err != nil {
-			return fmt.Errorf("failed to update hostedcluster after defaulting IDs: %w", err)
+		if !equality.Semantic.DeepEqual(orig, hcluster) {
```
Comment (alvaroaleman) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4279]: Remove the `needsUpdate` maybe?

**Comment by ironcladlou** [comment]
I want to revisit this in context of the createOrUpdate utility

**Pull request closed without merging**